File: kotlin-electron/karakum/plugins/convertUnknownRecordKeyType.js
import ts from "typescript";

export default function (node) {
    if (
        (
            node.kind === ts.SyntaxKind.UnknownKeyword
            || node.kind === ts.SyntaxKind.AnyKeyword
        )

        && node.parent
        && ts.isTypeReferenceNode(node.parent)
        && ts.isIdentifier(node.parent.typeName)
        && node.parent.typeName.text === "Record"

        && node === node.parent.typeArguments[0]
    ) {

        return "Any"
    }
    return null
}

File: kotlin-electron/karakum/src/jsMain/kotlin/electron/karakum/plugins/convertUnknownRecordKeyType.kt
package electron.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getParentOrNull
import typescript.SyntaxKind
import typescript.asArray
import typescript.isIdentifier
import typescript.isTypeReferenceNode

val convertUnknownRecordKeyType = createPlugin { node, _, _ ->
    nullable {
        ensure(
            node.kind == SyntaxKind.UnknownKeyword
                    || node.kind == SyntaxKind.AnyKeyword
        )

        val typeReference = ensureNotNull(node.getParentOrNull())
        ensure(isTypeReferenceNode(typeReference))

        val typeName = typeReference.typeName
        ensure(isIdentifier(typeName))
        ensure(typeName.text == "Record")

        ensure(node === typeReference.typeArguments?.asArray()?.firstOrNull())

        "Any"
    }
}
